events/nats_subscribe.go

```diff
@@ -0,0 +1,172 @@
+package events
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/nats-io/nats.go"
+)
+
+func (c *NatsConnection) coreSubscribe(ctx context.Context, subject string) (<-chan *nats.Msg, error) {
+	logger := c.logger.With(
+		"nats.provider", "core",
+		"nats.subject", subject,
+	)
+
+	sub, err := c.conn.SubscribeSync(subject)
+	if err != nil {
+		return nil, err
+	}
+
+	msgCh := make(chan *nats.Msg, *c.cfg.SubscriberFetchBatchSize)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				close(msgCh)
+
+				if err := sub.Unsubscribe(); err != nil {
+					logger.Warnw("error unsubscribing", "error", err)
+				}
+
+				return
+			default:
+			}
+
+			if err := c.nextMessage(ctx, sub, msgCh); err != nil {
+				if !errors.Is(err, context.DeadlineExceeded) {
+					logger.Errorw("error fetching messages", "error", err)
+				}
+
+				select {
+				case <-ctx.Done():
+				case <-time.After(*c.cfg.SubscriberFetchDelay):
+				}
+			}
+		}
+	}()
+
+	return msgCh, nil
+}
+
+func (c *NatsConnection) jsSubscribe(ctx context.Context, subject string) (<-chan *nats.Msg, error) {
+	durableName := c.durableName(subject)
+
+	logger := c.logger.With(
+		"nats.provider", "jetstream",
+		"nats.subject", subject,
+		"nats.durable_name", durableName,
+	)
+
+	sub, err := c.jetstream.PullSubscribe(subject, durableName, c.cfg.SubscribeOptions...)
+	if err != nil {
+		return nil, err
+	}
+
+	msgCh := make(chan *nats.Msg, *c.cfg.SubscriberFetchBatchSize)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				close(msgCh)
+
+				if err := sub.Unsubscribe(); err != nil {
+					logger.Warnw("error unsubscribing", "error", err)
+				}
+
+				return
+			default:
+			}
+
+			if err := c.fetchMessages(ctx, sub, msgCh); err != nil {
+				if !errors.Is(err, context.DeadlineExceeded) {
+					logger.Errorw("error fetching messages", "error", err)
+				}
+
+				select {
+				case <-ctx.Done():
+				case <-time.After(*c.cfg.SubscriberFetchDelay):
+				}
+			}
+		}
+	}()
+
+	return msgCh, nil
+}
+
+func (c *NatsConnection) fetchMessages(ctx context.Context, sub *nats.Subscription, msgCh chan<- *nats.Msg) error {
+	ctx, cancel := context.WithTimeout(ctx, *c.cfg.SubscriberFetchTimeout)
+
+	defer cancel()
+
+	batch, err := sub.FetchBatch(*c.cfg.SubscriberFetchBatchSize, nats.Context(ctx))
+	if err != nil {
+		return err
+	}
+
+	for msg := range batch.Messages() {
+		msgCh <- msg
```

Comment:
If the channel is unbuffered or the buffer is full, this could block indefinitely. It may be a good idea to use a `select` with a case for `msgCh <- msg` and a case for `<-ctx.Done()`. That way, the send operation will be canceled when the context finishes.

Reply:
good point, thanks!